wcfsetup/install/files/lib/data/option/OptionAction.class.php

```diff
@@ -111,4 +115,84 @@ public function emailSmtpTest() {
 		
 		return ['validationResult' => $smtp->testConnection()];
 	}
+	
+	/**
+	 * Validates the "generateRewriteRules" action
+	 */
+	public function validateGenerateRewriteRules() {
+		//TODO
+	}
+	
+	/**
+	 * Returns a list of code-bbcode-containers containing the necessary
+	 * rewrite rules
+	 *
+	 * @return string
+	 */
+	public function generateRewriteRules() {
+		$output = '';
+		$tpl = TemplateEngine::getInstance();
+		
+		foreach ($this->fetchRewriteRules() as $path => $content) {
+			$content = explode("\n", $content);
+			$output.= $tpl->fetch('codeMetaCode', 'wcf', [
+				'highlighter' => PlainHighlighter::getInstance(),
+				'filename' => $path,
+				'startLineNumber' => 1,
+				'content' => $content,
+				'lines' => count($content)
+			]);
+		}
+		
+		return $output;
+	}
+	
+	/**
+	 * Returns an array with rewrite rules per necessary directory/file
+	 * Applications in sub-directories of another application will be mapped to the top one
+	 *
+	 * @return string[]
+	 */
+	protected function fetchRewriteRules() {
+		$dirs = $rules = [];
+		foreach (ApplicationHandler::getInstance()->getApplications() as $app) {
+			$test = $app->getPackage()->getAbsolutePackageDir();
+			$insert = true;
+			
+			foreach ($dirs as $dir => $apps) {
+				if (strpos($dir, $test) !== false) {
+					unset($dirs[$dir]);
+					$insert = false;
+					break;
+				} else if (strpos($test, $dir) !== false) {
+					$insert = false;
+					break;
+				}
+			}
+			
+			if ($insert) $dirs[$test] = [];
+		}
+		
+		foreach (ApplicationHandler::getInstance()->getApplications() as $application) {
+			foreach ($dirs as $dir => $value) {
+				if (strpos($application->getPackage()->getAbsolutePackageDir(), $dir) !== false) {
+					$dirs[$dir][$application->domainPath] = $application->getPackage()->getAbsolutePackageDir();
+				}
+			}
+		}
+		
+		foreach ($dirs as $dir => $domainPaths) {
+			krsort($domainPaths);
+			
+			foreach ($domainPaths as $domainPath => $value) {
+				$path = FileUtil::removeTrailingSlash(substr($value, strlen($dir)));
+				$snippet = 'RewriteCond %{SCRIPT_FILENAME} !-d
```

Comment:
Consider using Heredoc instead, it makes the code much easier to read and puts an emphasis on the exact look of it:
```php
$snippet = <<<SNIPPET
[…]
SNIPPET;
```